### PR TITLE
[Merged by Bors] - feat: define `HasCountableSeparatingOn`

### PR DESCRIPTION
**Pull request by urkud**
- Define a typeclass saying that a countable family of sets satisfying
  a given predicate separate points of a given set.
- Provide instances for open and closed sets in a T₀ space with second
  countable topology and for measurable sets in case of a countably
  generated σ-algebra.

See [Zulip](https://leanprover.zulipchat.com/#narrow/stream/116395-maths/topic/ae.20eq.20of.20preimages) for motivation.

---
I'm not sure about the name.

- [x] depends on: #5636

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by semorrison** [comment]
This PR/issue depends on:
* leanprover-community/mathlib4#5636
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib4#5636~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

File: Mathlib/MeasureTheory/MeasurableSpace.lean [hunk 1]
```diff
@@ -1664,42 +1665,36 @@ instance [MeasurableSpace α] [CountablyGenerated α] [MeasurableSpace β] [Coun
...
-  refine' ⟨fun x n => x ∈ e n, _, _⟩
-  · rw [measurable_pi_iff]
+  rcases h.1 with ⟨S, hSc, hSm, hS⟩
+  wlog hne : S.Nonempty generalizing S
+  · exact this (insert ∅ S) (hSc.insert ∅) (forall_insert_of_forall hSm .empty)
+      (fun x hx y hy h ↦ hS x hx y hy (forall_of_forall_insert h)) (insert_nonempty _ _)
+  obtain ⟨e, rfl⟩ := Set.Countable.exists_eq_range hSc hne
+  simp only [forall_range_iff] at hSm hS
+  refine ⟨fun x n ↦ x ∈ e n, ?_, ?_⟩
+  . rw [measurable_pi_iff]
```
Comment (urkud) [inline Mathlib/MeasureTheory/MeasurableSpace.lean:1692]:
```suggestion
  · rw [measurable_pi_iff]
```

File: Mathlib/Order/Filter/CountableSeparatingOn.lean [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+This typeclass is used in all lemmas in this file to avoid repeating them for open sets, closed
+sets, and measurable sets.
+
+### Main results
+
+#### Filters supported on a (sub)singleton
+
+Let `l : Filter α` be a filter with countable intersections property. Let `p : Set α → Prop` be a
+property such that there exists a countable family of sets satisfying `p` and separating points of
+`α`. Then there `l` is supported on a subsingleton: there exists a subsingleton `t` such that
```
Comment (sgouezel) [inline Mathlib/Order/Filter/CountableSeparatingOn.lean:37]:
```suggestion
`α`. Then `l` is supported on a subsingleton: there exists a subsingleton `t` such that
```

File: Mathlib/Order/Filter/CountableSeparatingOn.lean [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+`Filter.exists_subset_subsingleton_mem_of_forall_separating`,
+`Filter.exists_mem_singleton_mem_of_mem_of_nonempty_of_forall_separating`,
+`Filter.exists_singleton_mem_of_mem_of_forall_separating`,
+`Filter.exists_subsingleton_mem_of_forall_separating`, and
+`Filter.exists_singleton_mem_of_forall_separating`.
+
+#### Eventually constant functions
+
+Consider a function `f : α → β`, a filter `l` with countable intersections property, and a countable
+separating family of sets of `β̀`. Suppose that for every `U` from the family, either
```
Comment (sgouezel) [inline Mathlib/Order/Filter/CountableSeparatingOn.lean:50]:
```suggestion
separating family of sets of `β`. Suppose that for every `U` from the family, either
```

File: Mathlib/Order/Filter/CountableSeparatingOn.lean [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+`∀ᶠ x in l, f x ∈ U` or `∀ᶠ x in l, f x ∉ U`. Then `f`` is eventually constant along `l`.
```
Comment (sgouezel) [inline Mathlib/Order/Filter/CountableSeparatingOn.lean:51]:
```suggestion
`∀ᶠ x in l, f x ∈ U` or `∀ᶠ x in l, f x ∉ U`. Then `f` is eventually constant along `l`.
```

File: Mathlib/Order/Filter/CountableSeparatingOn.lean [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+
+open Function Set Filter
+
+/-- We say that a type `α` has a *countable separating family of sets* satisfying a predicate
+`p : Set α → Prop` on a set `t` if there exists a countable family of sets `S : Set (Set α)` such
+that all sets `s ∈ S` satisfy `p` and any two distinct points `x y ∈ t`, `x ≠ y`, can be separated
+by `s ∈ S`.
+
+E.g., if `α` is a `T₀` topological space with second countable topology, then it has a countable
+separating family of open sets and a separating family of closed sets.
```
Comment (sgouezel) [inline Mathlib/Order/Filter/CountableSeparatingOn.lean:81]:
```suggestion
separating family of open sets and a countable separating family of closed sets.
```

File: Mathlib/Order/Filter/CountableSeparatingOn.lean [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+
+filter, countable
+-/
+
+open Function Set Filter
+
+/-- We say that a type `α` has a *countable separating family of sets* satisfying a predicate
+`p : Set α → Prop` on a set `t` if there exists a countable family of sets `S : Set (Set α)` such
+that all sets `s ∈ S` satisfy `p` and any two distinct points `x y ∈ t`, `x ≠ y`, can be separated
+by `s ∈ S`.
```
Comment (sgouezel) [inline Mathlib/Order/Filter/CountableSeparatingOn.lean:78]: Can you be more specific about what this means to be separated here? (it's not: "there exists `s` such that `x` belongs to `s` and `y` doesn't", its "there exists `s` such that `x` belongs to `s` and `y` doesn't or conversely")

File: Mathlib/Order/Filter/CountableSeparatingOn.lean [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+namespace Filter
+
+variable {l : Filter α} [CountableInterFilter l] {f g : α → β}
+
+/-!
+### Filters supported on a (sub)singleton
+
+In this section we prove several versions of the following theorem. Let `l : Filter α` be a filter
+with countable intersections property. Let `p : Set α → Prop` be a property such that there exists a
+countable family of sets satisfying `p` and separating points of `α`. Then there `l` is supported on
```
Comment (sgouezel) [inline Mathlib/Order/Filter/CountableSeparatingOn.lean:113]:
```suggestion
countable family of sets satisfying `p` and separating points of `α`. Then `l` is supported on
```

File: Mathlib/Order/Filter/CountableSeparatingOn.lean [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+a subsingleton: there exists a subsingleton `t` such that `t ∈ l`.
+
+With extra `Nonempty`/`Set.Nonempty` assumptions one can assume that `t` is a singleton `{x}`. If
+`s ∈ l`, then it suffices to assume that the countable family separates only points of `s`.
```
Comment (sgouezel) [inline Mathlib/Order/Filter/CountableSeparatingOn.lean:117]:
```suggestion
With extra `Nonempty`/`Set.Nonempty` assumptions one can ensure that `t` is a singleton `{x}`. 

If `s ∈ l`, then it suffices to assume that the countable family separates only points of `s`.
```

File: Mathlib/Order/Filter/CountableSeparatingOn.lean [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+theorem exists_singleton_mem_of_forall_separating [Nonempty α] (p : Set α → Prop)
+    [HasCountableSeparatingOn α p univ] (hl : ∀ U, p U → U ∈ l ∨ Uᶜ ∈ l) :
+    ∃ x : α, {x} ∈ l :=
+  exists_singleton_mem_of_mem_of_forall_separating p univ_mem hl
+
+/-!
+### Eventually constant functions
+
+In this section we apply theorems from the previous section to the filter `Filter.map f l` to show
+that `f : α → β`` is eventually constant along `l` if for every `U` from the separating family,
```
Comment (sgouezel) [inline Mathlib/Order/Filter/CountableSeparatingOn.lean:167]:
```suggestion
that `f : α → β` is eventually constant along `l` if for every `U` from the separating family,
```

File: Mathlib/Topology/CountableSeparatingOn.lean [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+second countable, then there exists a countable family of open sets in `X` that separates points
+of `s`. -/
+instance [TopologicalSpace X] {s : Set X} [T0Space s] [SecondCountableTopology s] :
+    HasCountableSeparatingOn X IsOpen s := by
+  suffices HasCountableSeparatingOn s IsOpen univ from .of_subtype fun _ ↦ isOpen_induced_iff.1
+  refine ⟨⟨countableBasis s, countable_countableBasis _, fun _ ↦ isOpen_of_mem_countableBasis,
+    fun x _ y _ h ↦ ?_⟩⟩
+  exact ((isBasis_countableBasis _).inseparable_iff.2 h).eq
+
+/-- If there exists a countbale family of open sets separating points of `s`, then there exists
```
Comment (sgouezel) [inline Mathlib/Topology/CountableSeparatingOn.lean:28]:
```suggestion
/-- If there exists a countable family of open sets separating points of `s`, then there exists
```

File: Mathlib/MeasureTheory/MeasurableSpace.lean [hunk 11]
```diff
@@ -1664,42 +1665,36 @@ instance [MeasurableSpace α] [CountablyGenerated α] [MeasurableSpace β] [Coun
...
+  rcases h.1 with ⟨b, hbc, hb⟩
+  refine ⟨⟨b, hbc, fun t ht ↦ hb.symm ▸ .basic t ht, fun x _ y _ h ↦ ?_⟩⟩
+  rw [← forall_generateFrom_mem_iff_mem_iff, ← hb] at h
+  simpa using h {y}
+
 variable (α)
 
 open Classical
 
 /-- If a measurable space is countably generated, it admits a measurable injection
```
Comment (sgouezel) [inline Mathlib/MeasureTheory/MeasurableSpace.lean:1680]:
```suggestion
/-- If a measurable space is countably generated and separates points, it admits a measurable injection
```

File: Mathlib/MeasureTheory/MeasurableSpace.lean [hunk 12]
```diff
@@ -1664,42 +1665,36 @@ instance [MeasurableSpace α] [CountablyGenerated α] [MeasurableSpace β] [Coun
...
+    [h : HasCountableSeparatingOn α MeasurableSet univ] :
     ∃ f : α → ℕ → Bool, Measurable f ∧ Function.Injective f := by
-  obtain ⟨b, bct, hb⟩ := h.isCountablyGenerated
-  obtain ⟨e, he⟩ := Set.Countable.exists_eq_range (bct.insert ∅) (insert_nonempty _ _)
-  rw [← generateFrom_insert_empty, he] at hb
-  refine' ⟨fun x n => x ∈ e n, _, _⟩
+  rcases h.1 with ⟨S, hSc, hSm, hS⟩
+  wlog hne : S.Nonempty generalizing S
+  · exact this (insert ∅ S) (hSc.insert ∅) (forall_insert_of_forall hSm .empty)
+      (fun x hx y hy h ↦ hS x hx y hy (forall_of_forall_insert h)) (insert_nonempty _ _)
```
Comment (sgouezel) [inline Mathlib/MeasureTheory/MeasurableSpace.lean:1688]: could you have this as a separate lemma in the `HasCountableSeparatingOn`: given a set `s` which satisfies `p s`, then there is a nonempty countable set `S` separating the points and satisfying `p`. And another lemma that gives this as the range of a function from nat?

**Comment by sgouezel** [comment]
bors r+
Thanks!

**Pull request closed without merging**